### ISSUE TITLE
Convert Stripe checkout form to use Stripe Elements

### ISSUE THREAD

**Issue by nosegraze** [restrictcontentpro/restrict-content-pro#1462]
From https://github.com/easydigitaldownloads/edd-stripe/issues/183 :

> Doc: https://stripe.com/docs/stripe.js#collecting-card-details
> 
> They are moving away from having custom forms for collecting card information and would like people to start using the Elements library to create the forms. This requires including their `v3` of the `stripe.js`.
>
> While v3 isn't fully featured yet, Stripe suggests loading both v2 and v3 at the same time until then.
>
> > This method of using Stripe.js to collect card information using a custom payment form has been **deprecated**. When creating a payment form to collect card information, use Stripe Elements—our pre-built UI components. Check out our Elements migration guide to learn how to migrate your checkout flow to Elements.

**Comment by chrisvanpatten** [comment]
At @tomodomoco we are working on a project that has Stripe Elements in the scope. We will likely need to do a custom job to hit the milestone but after that's done I'd be happy to direct the team to rework it as a PR for RCP core.

**Comment by mindctrl** [comment]
Related: https://github.com/restrictcontentpro/restrict-content-pro/issues/1227

**Comment by chrisvanpatten** [comment]
So we're just diving in on this in our client project (cc'ing @mcfarlan who will be working on this for us).

Theoretically it should be as easy as modifying the `fields()`, `scripts()`, and `validate_fields()` methods in `RCP_Payment_Gateway_Stripe` to use the Elements code, v3 script URL, and remove some unneeded validation. Ideally the options will be somewhat filterable so users can customize the styles themselves.

I think the main question would be in regards to backcompat.

In the interest of getting this live for our client in the next few days, we were planning to just register a new gateway that extends `RCP_Payment_Gateway_Stripe` as `RCP_Payment_Gateway_Stripe_Elements` (or something like that) and overrides the appropriate methods as mentioned above.

If we did the same for our pull request (Elements as a separate gateway), we could also rename the normal Stripe gateway as "Stripe (deprecated)" or "Stripe (legacy)" so users aren't too surprised by the change.

But I can also buy into the argument that since Stripe Elements comes somewhat pre-styled, it'd be better to transition everyone to it at once, as long as they receive ample warning.

Would love any and all thoughts so we implement this in the best way!

**Comment by pippinsplugins** [comment]
@chrisvanpatten I'd love to hear back on how well it works, at which time we can then decide how to handle merging it into core.

**Comment by chrisvanpatten** [comment]
We knocked out a V1 here: https://github.com/TomodomoCo/rcp-stripe-elements

In our testing it's working great! The main thing to note is that Stripe Elements isn't designed to style the _entire_ input — only the contents of the input. In other words, this is what you see by default:

![upgrade_ _ambient_institute](https://user-images.githubusercontent.com/1231306/31154911-53ddbff0-a878-11e7-9848-3aad6f49d07b.png)

The wrapper element (which we gave a class of `.rcp-stripe-elements-input` in our plugin) will need to have CSS applied that provides the border and any padding/margin for the "input", so you get something more like this:

![upgrade_ _ambient_institute](https://user-images.githubusercontent.com/1231306/31155256-b97a0452-a87a-11e7-9b5e-5b24049d7f4b.png)

@mcfarlan and I are happy to answer any questions on this!

**Comment by chrisvanpatten** [comment]
Update:

We did encounter an issue, which is that the `[rcp_update_card]` shortcode doesn't inherit the `fields()` from the active gateway. That means that the update card form shows up as the old style instead of the Elements form.

Right now we're going to fix it by rewriting `card-update-form.php` but this is probably not a good long term solution.

Here's a rough idea (please poke holes in this, I'm sure there are many): 

+ Gateways should be able to define `$this->supports[] = 'card-updates';` and this setting should be checked in `rcp_member_can_update_billing_card()` instead of the [current conditionals](https://github.com/restrictcontentpro/restrict-content-pro/blob/7d90a8565bfbf65690158870b5e45cb8b00de0cd/includes/member-functions.php#L1207-L1220).
+ If true, `[rcp_update_card]` would load, but the [code that defines the fields](https://github.com/restrictcontentpro/restrict-content-pro/blob/master/templates/card-update-form.php#L36-L69) in `card-update-form.php` would load `$current_gateway->update_fields()`. The method in the base gateway class would just load a new template containing the update fields as currently coded in the template (so it would preserve the current behaviour) or could override it with their own fields (as in the case of Stripe Elements).

**Comment by nosegraze** [comment]
Thanks for all your notes on this @chrisvanpatten :)

IMO we should start looking at bringing this into core soon-ish. Even though it doesn't look like we have to meet that November 1st deadline I think it's still a good idea to get started sooner rather than later. Stripe really hasn't handled this well at all and I don't want to be caught by surprise again.

**Comment by chrisvanpatten** [comment]
Agreed! Happy to contribute any/all of our Stripe Elements code. Just let me know what you need :)
